### PR TITLE
Implement settable timeouts in Functions.

### PR DESCRIPTION
**Pull request by bklimt**
1. Update okhttp to a newer version.
2. Change the default timeout for Functions to 60 seconds, to match the backend.
3. Add methods to change the timeout for a Function.
4. Ensure all exceptions coming from Functions are wrapped in FirebaseFunctionsException.

### CODE REVIEW

**Comment by bklimt** [comment]
These oss-bot failures don't look like they have anything to do with my changes.

**Comment by bklimt** [comment]
This was implemented with okhttp 3, because the newer version: 1) lets us specify a single timeout for the whole call, instead of separate timeouts for connect, read, and write, and 2) lets us easily set timeouts for particular calls, rather than having to create a new client for each callable.

https://github.com/square/okhttp/wiki/Recipes#timeouts

The downside is this may increase library size for developers using both Firestore and Functions, since they will no longer share the same okhttp version.

**Comment by CodingDoug** [comment]
Is 60s the best number on the client?  If that's measured by the moment the client attempts to connect, to the moment it receives the response, the function on the server may not actually have its own full 60s to yield a response, since the client could also be waiting for network latency and such.

**Comment by bklimt** [comment]
60s is what had been agreed upon for the API, and it's a big improvement over the previous, unspecified default. But I agree it would be reasonable to have a longer default to account for connection time. We should make the default the same across all client SDKs. What length default timeout would you propose?

**Comment by CodingDoug** [comment]
I'd guess that a buffer of 5 extra seconds is enough to cover 98% of the connections out there.

**Comment by bklimt** [comment]
After some discussion offline, I've upped the timeout to 70s. Thanks for the feedback @CodingDoug.

File: firebase-functions/firebase-functions.gradle [hunk 1]
```diff
@@ -55,7 +55,7 @@ dependencies {
     }
     implementation 'com.google.firebase:firebase-iid-interop:16.0.1'
 
-    implementation 'com.squareup.okhttp:okhttp:2.7.5'
+    implementation 'com.squareup.okhttp3:okhttp:3.12.1'
```
[the remark this answers is not in the document]
Comment (ashwinraghav) [reply firebase-functions/firebase-functions.gradle]: I might try to revive some of those bugs on grpc

File: firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallOptions.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.functions;
+
+import java.util.concurrent.TimeUnit;
+import okhttp3.OkHttpClient;
+
+/** An internal class for keeping track of options applied to an HttpsCallableReference. */
+class HttpsCallOptions {
```
Comment (ashwinraghav) [inline firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallOptions.java]: Question: Are there some functions that could take more time?
Should this be configurable/part of public API?

Comment (ashwinraghav) [reply firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallOptions.java]: NVM. I see it down there :)

File: firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallableReference.java [hunk 3]
```diff
@@ -89,6 +94,28 @@
...
+    return functionsClient.call(name, null, options);
+  }
+
+  /**
+   * Changes the timeout for calls from this instance of Functions. The default is 60 seconds.
+   *
+   * @param timeout The length of the timeout, in the given units.
+   * @param units The units for the specified timeout.
+   */
+  public void setTimeout(long timeout, TimeUnit units) {
```
Comment (ashwinraghav) [inline firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallableReference.java]: Does this require an API review?

File: firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java [hunk 4]
```diff
@@ -243,17 +244,28 @@ public void useFunctionsEmulator(String origin) {
       request = request.header("Firebase-Instance-ID-Token", context.getInstanceIdToken());
     }
 
+    OkHttpClient callClient = options.apply(client);
```
Comment (ashwinraghav) [inline firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java]: Question: Do we create a client and apply these options to create a new Client for some reason?
I am wondering whether we can reduce state by just saying options.apply(new OKHttpClient()) or possibly options.createClient()

File: firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallableReference.java (hunk 3 above)
Comment (bklimt) [reply firebase-functions/src/main/java/com/google/firebase/functions/HttpsCallableReference.java]: Yes, the API has been approved. go/callable-timeouts

File: firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java (hunk 4 above)
Comment (bklimt) [reply firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java]: The way `options.apply()` is implemented is the recommended way to do it. To change the timeout, you have to create a new client from the existing client, but they share most of their state, so it's not as expensive as it looks. The okhttp API is admittedly not obvious.

https://github.com/square/okhttp/wiki/Recipes#per-call-configuration

Comment (bklimt) [reply firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java]: As to why we don't just create a new one every time: doing it this way lets us share one connection pool and dispatcher across all of the function references for an app.

File: firebase-functions/firebase-functions.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (bklimt) [reply firebase-functions/firebase-functions.gradle]: > would you be open to picking that back up?

I'm open to it, but we'd have to discuss its priority relative to the other things I'm working on.

[the remark this answers is not in the document]
Comment (bklimt) [reply firebase-functions/firebase-functions.gradle]: From the comments on those grpc issues, it sounds like maybe they haven't considered the bloat from developers also using okhttp3 as a reason for them to update their dependency. Perhaps we could follow up and suggest that as a reason to reopen the issue.

File: firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java (hunk 4 above)
Comment (ashwinraghav) [reply firebase-functions/src/main/java/com/google/firebase/functions/FirebaseFunctions.java]: Thanks for clarifying

**Pull request merged**